man/yhttp.3
.\" Copyright (c) 2022 Emil Engler <engler+yhttp@unveil2.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd May 7, 2022
.Dt YHTTP 3
.Os
.Sh NAME
.Nm yhttp
.Nd minimal HTTP server library
.Sh LIBRARY
.Lb libyhttp
.Sh SYNOPSIS
.In sys/types.h
.In stdint.h
.In yhttp.h
.Sh DESCRIPTION
The
.Nm yhttp
library offers a quick but elegant way to setup an HTTP/1.1 server.
Interfacing applications generally work as follows:
.Bl -enum
.It
Create the
.Vt "struct yhttp *"
instance by calling
.Xr yhttp_init 3 .
.It
Call
.Xr yhttp_dispatch 3
with the just created
.Vt "struct yhttp *" ,
the callback function to be executed, whenever a valid request passes through,
and a
.Vt "void *"
being an argument of the previous callback function.
.It
Eventually stop the execution of
.Xr yhttp_dispatch 3
by calling
.Xr yhttp_stop 3 .
.It
Clean-up the resources by passing the
.Vt "struct yhttp *"
to
.Xr yhttp_free 3
and eventually exit the application.
.El
.Ss Pledge Promises
The
.Nm yhttp
library is built to operate in security-sensitive environments, including
.Xr pledge 2
on
.Ox .
The following is a list of all required promises.
.Bl -tag -width Ds
.It Va stdio
Required by essentially all
.Nm
functions.
.It Va inet
Required by
.Xr yhttp_dispatch 3 .
.El
.Sh EXAMPLES
The following sets up an HTTP/1.1 server on TCP port 8080, that replies with
a quote from
.Dq The Beatles
in the Content-Type
.Qq text/plain .
.Bd -literal -offset indent
#include <sys/types.h>

#include <err.h>
#include <signal.h>
#include <stdint.h>
#include <string.h>

#include <yhttp.h>

static void	cb(struct yhttp_requ *, void *);
static void	sighdlr(int);

static struct yhttp	*yh;

static void
cb(struct yhttp_requ *requ, void *unused)
{
	const char	*resp;

	resp = "Love has a nasty habit of disappearing overnight";

	/* Not required, because 200 is the default status code anyway. */
	yhttp_resp_status(requ, 200);
	yhttp_resp_header(requ, "Content-Type", "text/plain");
	yhttp_resp_body(requ, resp, strlen(resp));
}

static void
sighdlr(int sig)
{
	if (yhttp_stop(yh) != YHTTP_OK)
		errx(1, "yhttp_stop");
}

int
main(void)
{
	/* Setup the signal handlers for clean termination. */
	signal(SIGINT, sighdlr);
	signal(SIGTERM, sighdlr);

	/* Create the yhttp instance. */
	if ((yh = yhttp_init(8080)) == NULL)
		errx(1, "yhttp_init");

	/* Run the actual HTTP server. */
	if (yhttp_dispatch(yh, cb, NULL) != YHTTP_OK)
		errx(1, "yhttp_dispatch");

	/* Clean the resources up and exit the application. */
	yhttp_free(&yh);
	return (0);
}
.Ed
.Sh SEE ALSO
.Xr yhttp_dispatch 3 ,
.Xr yhttp_header 3 ,
.Xr yhttp_init 3 ,
.Xr yhttp_resp_status 3 ,
.Xr yhttp_url_enc 3
.Sh STANDARDS
Many standards are involved in the
.Nm
library, most significantly being RFC 7320
.Dq Hypertext Transfer Protocol (HTTP/1.1): Message Syntax and Routing .
.Sh AUTHORS
The
.Nm
library was written by
.An Emil Engler Aq Mt engler+yhttp@unveil2.org .
.Sh BUGS
Yes.
.Sh SECURITY CONSIDERATIONS
The usage of TCP ports below 1024 is forbidden, because it would require root
privileges, which is discouraged.
If you demand to use such a port, consider redirecting the traffic through
another tool, such as
.Xr pf 4 .
.Pp
A TLS implementation is not being provided on purpose.
If you demand to use HTTPS, please tunnel the connections over a tool such
as
.Xr stunnel 8
or
.Xr relayd 8 .
.Pp
It is strongly advised to run yhttp in a sandbox, such as
.Xr pledge 2
on
.Ox ,
due to the fact of being a network facing application.


CHANGES
1.0 (2022-05-07):
-----------------
- Initial release.


yhttp.c
/*
 * Copyright (c) 2022 Emil Engler <engler+yhttp@unveil2.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>

#include <assert.h>
#include <ctype.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "abnf.h"
#include "hash.h"
#include "yhttp.h"
#include "yhttp-internal.h"
#include "net.h"

/*
 * Functions from yhttp.h.
 */

struct yhttp *
yhttp_init(uint16_t port)
{
	struct yhttp	*yh;

	/*
	 * Running on a well-known port requires running with root privileges,
	 * which is strongly discouraged.  Therefore, only ports outside of
	 * the well-known range are permitted for usage.
	 */
	if (port < 1024)
		return (NULL);

	if ((yh = malloc(sizeof(struct yhttp))) == NULL)
		return (NULL);

	memset(yh->pipe, -1, sizeof(yh->pipe));
	yh->is_dispatched = 0;
	yh->port = port;

	return (yh);
}

void
yhttp_free(struct yhttp **yh)
{
	if (yh == NULL || *yh == NULL)
		return;

	free(*yh);
	*yh = NULL;
}

char *
yhttp_header(struct yhttp_requ *requ, const char *name)
{
	struct yhttp_requ_internal	*internal;
	struct hash			*node;

	internal = requ->internal;
	if ((node = hash_get(internal->headers, name)) == NULL)
		return (NULL);
	else
		return (node->value);
}

char *
yhttp_query(struct yhttp_requ *requ, const char *key)
{
	struct yhttp_requ_internal	*internal;
	struct hash			*node;

	internal = requ->internal;
	if ((node = hash_get(internal->queries, key)) == NULL)
		return (NULL);
	else
		return (node->value);
}


/*
 * The following function is largely based upon kcgi(3)s khttp_urlencode(),
 * and is distributed under the following license:
 *
 * Copyright (c) 2022 Emil Engler <engler+yhttp@unveil2.org>
 * Copyright (c) 2012, 2014--2017 Kristaps Dzonsons <kristaps@bsd.lv>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
char *
yhttp_url_enc(const char *s)
{
	char	*res;
	char	 ch;
	size_t	 sz, i;
	int	 rc;

	if (s == NULL)
		return (NULL);

	/*
	 * In the worst case, the resulting string will be three times larger
	 * than the original, because every percent-encoded sequence is three
	 * bytes long.  Instead of performing several calls to realloc(3), we
	 * assume the worst case and simply make the resulting string three
	 * times larger, so we do not have to worry about any reallocations.
	 */

	/* Check for multiplication overflow. */
	sz = strlen(s) + 1;
	if (sz > SIZE_MAX / 3)
		return (NULL);

	if ((res = malloc(sz * 3)) == NULL)
		return (NULL);

	i = 0;
	while ((ch = *s++) != '\0') {
		if (isalnum((unsigned char)ch) || ch == '-' || ch == '_' ||
		    ch == '.' || ch == '~') {
			res[i++] = ch;
			continue;
		}
		else if (ch == ' ') {
			res[i++] = '+';
			continue;
		}

		/* Convert byte to percent-encoded sequence. */
		rc = snprintf(res + i, 4, "%%%.2hhX", (unsigned char)ch);
		if (rc != 3) {
			free(res);
			return (NULL);
		}
		i += 3;
	}
	res[i] = '\0';

	return (res);
}

char *
yhttp_url_dec(const char *s)
{
	char	*res;
	char	 tmp[3];
	char	 ch;
	long	 val;
	size_t	 i;

	if (s == NULL)
		return (NULL);

	/* The result cannot be bigger than the original string. */
	if ((res = malloc(strlen(s) + 1)) == NULL)
		return (NULL);

	i = 0;
	while ((ch = *s) != '\0') {
		if (ch == '%') {
			/*
			 * Check if this is actually a valid and legal
			 * percent-encoded sequence, by checking if the
			 * following two bytes are valid hexadecimal
			 * characters.
			 */
			if (!isxdigit(s[1]) || !isxdigit(s[2]))
				goto err;

			/* Convert the ASCII chars into an integer. */
			strlcpy(tmp, s + 1, sizeof(tmp));
			val = strtol(tmp, NULL, 16);
			assert(val >= 0x0 && val <= 0xFF);

			/* %00 is forbidden for security reasons. */
			if (val == 0)
				goto err;

			/* Append it to the result. */
			res[i++] = val;
			s += 3;
			continue;
		} else if (ch == '+')
			res[i++] = ' ';
		else
			res[i++] = ch;
		++s;
	}
	res[i] = '\0';

	return (res);
err:
	free(res);
	return (NULL);
}

int
yhttp_resp_status(struct yhttp_requ *requ, int status)
{
	struct yhttp_requ_internal	*internal;

	if (requ == NULL || status < 100 || status > 999)
		return (YHTTP_EINVAL);

	internal = requ->internal;
	internal->resp->status = status;

	return (YHTTP_OK);
}

int
yhttp_resp_header(struct yhttp_requ *requ, const char *name, const char *value)
{
	struct yhttp_requ_internal	*internal;
	size_t				 i;

	if (requ == NULL || name == NULL)
		return (YHTTP_EINVAL);

	internal = requ->internal;

	if (value == NULL || *value == '\0') {
		/* Delete a previously set header field. */

		hash_unset(internal->resp->headers, name);
		return (YHTTP_OK);
	} else {
		/* Set a header field. */

		/* Validate name. */
		if (*name == '\0')
			return (YHTTP_EINVAL);
		for (i = 0; name[i] != '\0'; ++i) {
			if (!abnf_is_tchar(name[i]))
				return (YHTTP_EINVAL);
		}
		/* Transfer-Encoding and Content-Length may not be set. */
		if (strcasecmp(name, "Transfer-Encoding") == 0 ||
		    strcasecmp(name, "Content-Length") == 0)
			return (YHTTP_EINVAL);

		/* Validate value. */
		for (i = 0; value[i] != '\0'; ++i) {
			if (!isprint(value[i]))
				return (YHTTP_EINVAL);
		}

		return (hash_set(internal->resp->headers, name, value));
	}
}

int
yhttp_resp_body(struct yhttp_requ *requ, const unsigned char *body,
		size_t nbody)
{
	struct yhttp_requ_internal	*internal;

	if (requ == NULL)
		return (YHTTP_EINVAL);

	internal = requ->internal;
	free(internal->resp->body);

	if (body == NULL || nbody == 0) {
		/* Unset the message body. */

		internal->resp->body = NULL;
		internal->resp->nbody = 0;
		return (YHTTP_OK);
	} else {
		/* Set the message body. */

		if ((internal->resp->body = malloc(nbody)) == NULL)
			return (YHTTP_ERRNO);
		memcpy(internal->resp->body, body, nbody);
		internal->resp->nbody = nbody;
		return (YHTTP_OK);
	}
}

int
yhttp_dispatch(struct yhttp *yh, void (*cb)(struct yhttp_requ *, void *),
	       void *udata)
{
	int	rc;

	if (yh == NULL || cb == NULL)
		return (YHTTP_EINVAL);

	/* The instance has already been dispatched. */
	if (yh->is_dispatched)
		return (YHTTP_EBUSY);
	yh->is_dispatched = 1;

	/*
	 * We are going to create a pipe(2) for the server, so that shutdown
	 * can be communicated somehow (by sending EOF through the pipe(2)).
	 */
	if (pipe(yh->pipe) == -1)
		return (YHTTP_ERRNO);

	rc = net_dispatch(yh->port, yh->pipe[0], cb, udata);

	/* Destroy the pipe. */
	close(yh->pipe[0]);
	close(yh->pipe[1]);
	memset(yh->pipe, -1, sizeof(yh->pipe));

	yh->is_dispatched = 0;

	return (rc);
}

int
yhttp_stop(struct yhttp *yh)
{
	if (yh == NULL)
		return (YHTTP_EINVAL);
	if (!yh->is_dispatched)
		return (YHTTP_ENOENT);

	/* We tell the server to shutdown by closing its pipe(2). */
	close(yh->pipe[1]);
	yh->pipe[1] = -1;

	return (YHTTP_OK);
}

/*
 * Functions from yhttp-internal.h.
 */

struct yhttp_requ *
yhttp_requ_init(void)
{
	struct yhttp_requ_internal	*internal;
	struct yhttp_requ		*requ;

	if ((requ = malloc(sizeof(struct yhttp_requ))) == NULL)
		return (NULL);

	requ->path = NULL;
	requ->client_ip = NULL;

	/*
	 * The request body will be stored inside the buffer of the parser
	 * rather than in a separately allocated space, meaning the clean-up
	 * is not in our scope.
	 */
	requ->body = NULL;
	requ->nbody = 0;

	requ->method = YHTTP_GET;

	/* Initialize the internal field. */
	if ((internal = malloc(sizeof(struct yhttp_requ_internal))) == NULL)
		goto err;
	requ->internal = internal;

	internal->headers = NULL;
	internal->queries = NULL;
	internal->resp = NULL;

	if ((internal->headers = hash_init()) == NULL)
		goto err;
	if ((internal->queries = hash_init()) == NULL)
		goto err;
	if ((internal->resp = yhttp_resp_init()) == NULL)
		goto err;

	return (requ);
err:
	free(requ);
	if (internal != NULL) {
		hash_free(internal->headers);
		hash_free(internal->queries);
		yhttp_resp_free(internal->resp);
		free(internal);
	}
	return (NULL);
}

void
yhttp_requ_free(struct yhttp_requ *requ)
{
	struct yhttp_requ_internal	*internal;

	if (requ == NULL)
		return;

	internal = requ->internal;

	/* See the comment regarding requ->body inside yhttp_requ_init(). */
	free(requ->path);
	free(requ->client_ip);
	free(requ);

	hash_free(internal->headers);
	hash_free(internal->queries);
	yhttp_resp_free(internal->resp);
	free(internal);
}

struct yhttp_resp *
yhttp_resp_init(void)
{
	struct yhttp_resp	*resp;

	if ((resp = malloc(sizeof(struct yhttp_resp))) == NULL)
		return (NULL);

	if ((resp->headers = hash_init()) == NULL) {
		free(resp);
		return (NULL);
	}

	resp->body = NULL;
	resp->nbody = 0;
	resp->status = 200;

	return (resp);
}

void
yhttp_resp_free(struct yhttp_resp *resp)
{
	if (resp == NULL)
		return;

	hash_free(resp->headers);
	free(resp->body);
	free(resp);
}
